### PR TITLE
Introduce Typescript as the default language for the server

### PR DESCRIPTION
**Pull request by liangyuanruo**
## Problem

In order to demonstrate software design patterns, we must first decide on a production language and have a basic server running. Additionally, we should take advantage of Github features and add templates for PRs, issues etc.

## Solution

Javascript is a loosely-typed language. While this can be convenient, building a major project on top of a loosely-typed language can lead to hours of lost time debugging or production rollbacks due to server crashes. Using a statically-typed language such as Typescript will help developers catch mistakes early in the development process, and provide an enhanced IDE experience when inspecting types.

This PR sets up the necessary tooling so that developers can get straight into the code without hassle.

**Features**:

- Hot reloading during development
- Pre-configured build configuration for Typescript transpilation to ES5
- Markdown templates for
  - pull requests
  - bug reports
  - feature requests
  - technical specifications

**Improvements**

- ESLint configuration is now compatible with Typescript

**New dependencies**:
- `express`: Server framework

**New dev dependencies**:

- `nodemon` : Hot reloading
- `ts-node`: Typescript interpreter
- `typescript`: Typescript transpiler
- `@typescript-eslint/parser`: Typescript ESLint parser
- `@types/express`: Express types from DefinitelyTyped

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,22 +7,30 @@
     "test": "echo \"no test specified\"",
     "cz": "git-cz",
     "lint": "eslint .",
-    "lint:fix": "eslint --fix ."
+    "lint:fix": "eslint --fix .",
+    "dev": "nodemon --exec 'ts-node --files' src/server/index.ts",
```
Comment (LoneRifle) [inline package.json:11]: do you think it would be helpful to enable the debugger here?

**Comment by LoneRifle** [comment]
Also if you could, adjust the README appropriately

File: .eslintrc.js [hunk 2]
```diff
@@ -4,11 +4,14 @@ module.exports = {
     node: true,
   },
   extends: ['eslint:recommended', 'plugin:prettier/recommended'],
```
Comment (LoneRifle) [inline .eslintrc.js:6]: Should this be `plugin:@typescript-eslint/recommended`?

File: .eslintrc.js [hunk 3]
```diff
@@ -4,11 +4,14 @@ module.exports = {
...
+  parser: '@typescript-eslint/parser',
   parserOptions: {
     ecmaVersion: 2018,
+    sourceType: 'module',
   },
   rules: {
     'no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
     'no-console': 'warn',
```
Comment (LoneRifle) [inline .eslintrc.js:14]:
```suggestion
    'no-console': 'warn',
    '@typescript-eslint/member-delimiter-style': [
      'error',
      {
        multiline: {
          delimiter: 'none',
          requireLast: true,
        },
        singleline: {
          delimiter: 'semi',
          requireLast: false,
        },
      },
    ],
```

File: package.json (hunk 1 above)
Comment (liangyuanruo) [reply package.json:11]: is this the `--inspect` flag?

Comment (LoneRifle) [reply package.json:11]: yup, though I don't know how the presence of nodemon and ts-node would influence this

File: .eslintrc.js (hunk 3 above)
Comment (liangyuanruo) [reply .eslintrc.js:14]: good suggestion

**Pull request merged**